File: ai_module/src/task_planner/scripts/task_planner_client.py
import rospy

from task_planner.msg import Plan, Node, Edge, Task
from task_planner.srv import GetSubplans, GetSubplansResponse

def call_service():
    rospy.init_node('task_planner_client')
    rospy.wait_for_service('/task_planner/request_subplans')

    try:
        get_subplans = rospy.ServiceProxy('/task_planner/request_subplans', GetSubplans)
        # TODO: question should be inputted by users.
        question = "Count the number of pillows on the sofa."
        response = get_subplans(question)

        rospy.loginfo(response.plan)

    except rospy.ServiceException as e:
        rospy.logerr(f"Service call failed: {e}")

if __name__ == "__main__":
    call_service()

File: ai_module/src/task_planner/src/task_planner_client.cpp
#include <ros/ros.h>
#include <task_planner/GetSubplans.h>
#include <task_planner/Entity.h>
#include <task_planner/EntityRelation.h>
#include <task_planner/SubTask.h>

void callTaskPlannerService(int argc, char **argv) {
    ros::init(argc, argv, "task_planner_client");
    ros::NodeHandle nh;
    ros::service::waitForService("get_subplans");

    ros::ServiceClient client = nh.serviceClient<task_planner::GetSubplans>("get_subplans");
    task_planner::GetSubplans srv;
    // TODO: question should be inputted by users.
    srv.request.question = "TODO: question should be inputted by users.";

    if (client.call(srv)) {
        ROS_INFO("Question: %s", srv.request.question.c_str());
        ROS_INFO("[Constraints]");
        for (const auto &constraint : srv.response.plan.constraints) {
            ROS_INFO("--- action : %s ---", constraint.action.c_str());
            ROS_INFO("  entity : %s", constraint.entity.target_name.c_str());
        }
        ROS_INFO("[Steps]");
        for (const auto &step : srv.response.plan.steps) {
            ROS_INFO("--- action : %s ---", step.action.c_str());
            ROS_INFO("  entity : %s", step.entity.target_name.c_str());
        }
    } else {
        ROS_ERROR("Service call failed");
    }
}

int main(int argc, char **argv) {
    callTaskPlannerService(argc, argv);
    return 0;
}
